### PR TITLE
s4702860 ADNI Siamese Classifier 

### PR DESCRIPTION
**Pull request by lambdarm**
Pull request for classification problem of ADNI dataset based upon a siamese network. Three input siamese neural network is created and trained on ADNI dataset, these embeddings are then being used in order to train the classifier. 

Created 4 new .py format files:
    modules.py - handles all neural networks (siamese, classification)
    train.py - handles the training of the neural networks on new, unseen data, and accuracy curves
    predict.py - outputs UMAP class plots and accuracy information

Updated README.md
    includes overview of the problem space, example output, dependencies.

### CODE REVIEW

**Comment by lambdarm** [comment]
To the marker,

I have the current number of data loaded set to 1000 which takes about 5 minutes to LOAD the data, not even train it, The classifier only predicts one class (couldnt get it working),

I figure there is no point in you wasting time loading the data that isn't even going go work, 
So as to not to waste your time marking, in train.py, change the num value in the load_dataset call in main to 100 and 25 for train and validation respectively, I have 1000 in the commit as it is the final model that I used to supply the outputs obtained in the readme.

I'm 99% sure the issue is due to not having enough data to classify the validation set, as I'm only taking 1000 samples out of 20000, 

TLDR: change num value in train.py to 100 and 25 for train and test sets respectively in main, to not waste time loading large data for classifier that doesn't work.
Thanks and sorry for the hastle,
Reuben!

EDIT: I Found my error, in train.py i loaded the classifier on the triples, not the outputs of the embeddings. Annoying but what can you do. :(

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Good
- clear flow of information
- sufficient details on algorith, training choices, and usage
- relevant plots provided

Commit messages: Very poor

Code: 
- good design
- sufficient comments
- algorithm seem to be correctly implemented

Functionality/Performance: 
- **no final average performance numbers provided for either the siamese or classifier**
- looking at the provided plots, both networks seem to be performing very poorly

General comments:
- algorithm seems to solve the problem apporpriately, **although performance is somewhat unknown**

**Comment by lambdarm** [comment]
Hi Natasha, 
Thanks for inspecting my code! Just wanted to clarify regarding your comment saying no performance accuracy is given:

- task is to get 80% on testing accuracy using siamese
- In predict.py, there is a print output for the test accuracy of the classifier. 
- is this not enough?

Regarding the low accuracy:
- found my error, trained classifier on ADNI set instead of the embeddings of siamese 

Finally, regarding the commit messages:
- the messages i did, for what ever reason, shows when you press the three dots next to the message name
- guide i was following said have a "message name", which was the "updated x"
- following that, said include what you did in the additional section
- propably not the format you guys wanted, sorry about that!

Thanks!

**Comment by nathasha-naranpanawa** [comment]
Hello,

No problem!

Regarding the test accuracy, it would have been best to include the test set performance in the readme as we might not run code at the initial inspection. And judging by the validation accuracy in the plot, I doubt that your test set achieves >80% classification accuracy. Does it?

> found my error, trained classifier on ADNI set instead of the embeddings of siamese

Not sure what you mean by this. The correct algorithm would be to train the classifier on the embeddings generated by the trained siamese.

**Comment by lambdarm** [comment]
Ahh oh sorry about that, 
Yea the test accuracy was only 50%, which it was only ever predicting the one class (I believe 1).

Also, regarding the training on the ADNI set, I know I was meant to train on the embeddings, however I trained oj the set by accident.

Reason was due to my just copy and pasting the train line from the siamese, changing the model name, but forgetting the data name lol

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem partially, no classifier -3
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage, missing results for performance -2
Module present
Commenting minimal -1
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages sometimes -1
Progressive commits used
## Documentation
ReadMe acceptable, no refs -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
